src/convert.jl

```diff
@@ -0,0 +1,229 @@
+module Conversion
+
+    using ..SAM
+    using ..BAM
+
+    using BioSequences, BioAlignments
+    import Base.convert
+
+    global  const op_to_code = Dict(
+        'M' => 0x00,
+        'I' => 0x01,
+        'D' => 0x02,
+        'N' => 0x03,
+        'S' => 0x04,
+        'H' => 0x05,
+        'P' => 0x06,
+        '=' => 0x07,
+        'X' => 0x08,
+        'B' => 0x09,
+        '0' => 0x0a
+    )
+
+    function parse_cigar(record::SAM.Record)
+        
+        cigar_ops = ('M','I','D','N','S','H','P','=','X')
+        out = NamedTuple{(:L, :op), Tuple{Int64, Char}}[]
+
+        cigar = SAM.cigar(record)
+        start = 1
+
+        for (i,c) in enumerate(cigar)
+            if c ∈ cigar_ops
+                L = parse(Int, cigar[start:i-1])
+                op = c
+                push!(out, (L = L, op = op))
+                start = i+1
+            end
+        end
+        out
+    end
+
+    function convert(::Type{BAM.Record}, record::SAM.Record; header = nothing)
+
+        bam_record = BAM.Record()
+        set_fixed_length_fields!(bam_record, record, header)
+        set_variable_length_fields!(bam_record, record)
+
+        bam_record
+    end
+
+    function refname_to_refid(refname::String, header)
+        isnothing(header) && return -1
+        refname == "*" && return -1
+        
+        refs = findall(header, "SQ")
+        for (i,ref) in enumerate(refs)
+            ref["SN"] == refname && return i-1
+        end
+        error("Couldn't find reference $(refname) in header.")
+    end
+
+    """
+        This field is set as '*' when the information is unavailable, and set as '=' if RNEXT is identical RNAME
+    """
+    function get_nextrefname(record::SAM.Record)
+        !SAM.hasnextrefname(record) && return "*"
+        SAM.nextrefname(record) == "=" ? SAM.refname(record) : SAM.nextrefname(record)
+    end
+
+    function get_refname(record::SAM.Record)
+        !SAM.hasrefname(record) && return "*"
+        SAM.refname(record)
+    end
+
+    function set_fixed_length_fields!(bam_record::BAM.Record, record::SAM.Record, header)
+
+        cigar_ops = ('M','I','D','N','S','H','P','=','X')
+        n_cigar_op = sum(x ∈ cigar_ops for x in SAM.cigar(record); init = 0)
+
+        if n_cigar_op > typemax(UInt16) 
+            error("Cigar contains more than $(typemax(UInt16)) operations. This is currently unsupported. See section 4.2.2 of SAM/BAM specifications.")
+        end
+
+        bin = reg2bin(SAM.position(record)-1, SAM.alignlength(record))
+        
+        bam_record.refid = refname_to_refid(get_refname(record), header)
+        bam_record.pos = SAM.position(record)-1
+        bam_record.l_read_name = SAM.tempname(record) |> x -> length(x)+1 # +1 for NULL
+        bam_record.mapq = SAM.mappingquality(record)
+        bam_record.bin = bin
+        bam_record.n_cigar_op = n_cigar_op
+        bam_record.flag = SAM.flag(record)
```

Comment:
This should be .flags and not .flag twice.

Reply:
Done.